### PR TITLE
Add support for new fields in MISSION_CURRENT message

### PR DESCRIPTION
**Pull request by peterbarker**
- mission state (useful for "complete")
 - number of commands
 

@auturgy this might be problematic in that I've implemented the spec as-written which is the number of commands - but we consider the home location a command, so your current squence number is never going to equal this new "total" field.

### CODE REVIEW

**Comment by auturgy** [comment]
The waypoint 0 / home thing is a little problematic (in this context: mostly it is actually good to have it as a waypoint). I've been thinking about whether it's worth burning a capability bit on that issue, to support both styles of reporting/recording home.

**Comment by peterbarker** [comment]
> The waypoint 0 / home thing is a little problematic (in this context: mostly it is actually good to have it as a waypoint). I've been thinking about whether it's worth burning a capability bit on that issue, to support both styles of reporting/recording home.

I'd like to transition away from having the home location as a waypoint :-)  But, separate issue.

I think if we can change the wording on this message to "maximum expected waypoint number" rather than "total number of mission items" (regardless of the name), we can just send the highest sequence number in this packet.

**Comment by magicrub** [comment]
And here's the PR for the mavlink change https://github.com/ArduPilot/mavlink/pull/271

**Comment by peterbarker** [comment]
Oh, @hamishwillee on that total-number-of-waypoints thing.

**Comment by peterbarker** [comment]
> And here's the PR for the mavlink change [ArduPilot/mavlink#271](https://github.com/ArduPilot/mavlink/pull/271)

More likely this one: https://github.com/ArduPilot/mavlink/pull/306 :-)

**Comment by peterbarker** [comment]
I've adjusted this PR on the assumption that we'll change the wording on the mavlink message to exclude the home point from "total".

I've also allowed Copter to say it is paused when it is.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -613,7 +613,44 @@ void GCS_MAVLINK::handle_mission_request(const mavlink_message_t &msg)
...
+        // exclude home location from the count; see message definition.
+        num_commands -= 1;
+    }
+
+    mavlink_msg_mission_current_send(
+        chan,
+        seq,
+        num_commands, // total
+        mission_state(mission), // mission_state
+        0); // mission_mode (0 means unknown, need vehicle callback for this)
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:651]: could fill this in? current_mode_requires_mission()

**Review by tridge — CHANGES_REQUESTED** [review]
ok to merge once in_auto fixed

**Comment by hamishwillee** [comment]
I've added note to the upstream mavlink correction. In essence, yes - on the last mission item `sequence === total`. 

My bad - while I know that ardupilot keeps home location is the 0 position I don't think of it as a mission item. Any objection to me merging that update (or my slight proposed amendment)?

**Comment by peterbarker** [comment]
I've merged the associated mavlink change after updating it from mavlink/mavlink.

I've added setting of the mode field based on whether the vehicle is in a mode which requires the mission (pre-existing method to do that for AP_Arming).

**Pull request merged**